From 47ae2a8d95627854d4ad0e2b5bc5ac03f9ba2dfe Mon Sep 17 00:00:00 2001
From: Ferdinand Xu <cheng.a.xu@intel.com>
Date: Mon, 22 May 2017 13:40:06 +0800
Subject: [PATCH 1134/1164] CDH-54159 HIVE-16693: beeline "source" command
 freezes if you have a comment in it? (Ferdinand
 Xu, reviewed by Carter Shanklin and Xuefu Zhang)

(cherry picked from commit 1e086f2691fe58e23a43b92f6851aa824ac8d440)

Change-Id: Ie4706ca73dcf25c7a2d303bdd2657b88184e043c
---
 .../src/java/org/apache/hive/beeline/Commands.java |    6 ++----
 .../org/apache/hive/beeline/cli/TestHiveCli.java   |    9 +++++++++
 2 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/beeline/src/java/org/apache/hive/beeline/Commands.java b/beeline/src/java/org/apache/hive/beeline/Commands.java
index 612bda5..6d1e2fb 100644
--- a/beeline/src/java/org/apache/hive/beeline/Commands.java
+++ b/beeline/src/java/org/apache/hive/beeline/Commands.java
@@ -880,9 +880,8 @@ private boolean sourceFileInternal(File sourceFile) throws IOException {
     BufferedReader reader = null;
     try {
       reader = new BufferedReader(new FileReader(sourceFile));
-      String extra = reader.readLine();
-      String lines = null;
-      while (extra != null) {
+      String lines = null, extra;
+      while ((extra = reader.readLine()) != null) {
         if (beeLine.isComment(extra)) {
           continue;
         }
@@ -891,7 +890,6 @@ private boolean sourceFileInternal(File sourceFile) throws IOException {
         } else {
           lines += "\n" + extra;
         }
-        extra = reader.readLine();
       }
       String[] cmds = lines.split(";");
       for (String c : cmds) {
diff --git a/beeline/src/test/org/apache/hive/beeline/cli/TestHiveCli.java b/beeline/src/test/org/apache/hive/beeline/cli/TestHiveCli.java
index 9f42512..a53a7b4 100644
--- a/beeline/src/test/org/apache/hive/beeline/cli/TestHiveCli.java
+++ b/beeline/src/test/org/apache/hive/beeline/cli/TestHiveCli.java
@@ -47,6 +47,7 @@
   private final static String SOURCE_CONTEXT3 =
       "create table if not exists test.testSrcTbl3(sc3 string);";
   private final static String SOURCE_CONTEXT4 = "show tables;!ls;show tables;\nquit;";
+  private final static String SOURCE_CONTEXT5 = "-- test;\n;show tables;\nquit;";
   final static String CMD =
       "create database if not exists test;\ncreate table if not exists test.testTbl(a string, b "
           + "string);\n";
@@ -165,6 +166,14 @@ public void testSourceCmd3() {
   }
 
   @Test
+  public void testSourceCmd4() {
+    File f = generateTmpFile(SOURCE_CONTEXT5);
+    verifyCMD("source " + f.getPath() + ";", "testtbl", os,
+      new String[] { "--database", "test" }, ERRNO_OK, true);
+    f.delete();
+  }
+
+  @Test
   public void testSqlFromCmd() {
     verifyCMD(null, "", os, new String[] { "-e", "show databases;" }, ERRNO_OK, true);
   }
-- 
1.7.9.5

